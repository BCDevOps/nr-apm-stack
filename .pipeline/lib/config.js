'use strict';
process.on('unhandledRejection', (reason) => {
  console.log(reason);
  process.exit(1);
});

module.exports = class {
  constructor(options) {
    this.options = options
  }
  build() {
    const options = this.options
    const name = 'nress'
    //const phases = ['build', 'dev', 'test', 'prod']
    const changeId = options.pr //aka pull-request

    const phases0 = {
      transient: {build: true                          , dev: true                                                            , test: false                                                         , prod: false},
      name:      {build: `${name}`                     , dev: `${name}`                                                       , test: `${name}`                                                     , prod: `${name}`},
      suffix:    {build: `-build-${changeId}`          , dev: `-dev-${changeId}`                                              , test: `-test`                                                       , prod: `-prod`},
      phase:     {build: 'build'                       , dev: 'dev'                                                           , test: 'test'                                                        , prod: 'prod'},
      keycloak:  {
        build: null,
        dev: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
        test: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
        prod: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
      },
    };
    const phases = {};
    // Pivot configuration table, so that `phase name` becomes a top-level property
    // { namespace: { build: '-tools',  dev: '-dev'}}   ->  { build: { namespace: '-tools' }, dev: { namespace: '-dev' } }
    Object.keys(phases0).forEach((properyName) => {
      const property = phases0[properyName];
      Object.keys(property).forEach((phaseName) => {
        phases[phaseName] = phases[phaseName] || {};
        phases[phaseName][properyName] = property[phaseName];
      });
    });

    return {phases, options, environments: ['dev', 'test', 'prod']}
  }
}