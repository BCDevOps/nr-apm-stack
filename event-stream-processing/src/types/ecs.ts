import { Event } from './event';

export interface esc {
  Event: Event;
}
