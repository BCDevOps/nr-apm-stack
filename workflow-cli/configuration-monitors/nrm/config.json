{
  "type": "onpremise-monitor",
  "prefix": "nrids_agent_fluentbit_"
}