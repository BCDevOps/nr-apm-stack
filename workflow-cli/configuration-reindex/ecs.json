{
  "source": {
    "indexPrefix": "nrm-logs-access-"
  },
  "dest": {
    "indexPrefix": "nrm-access-external-"
  },
  "prototype": {
    "conflicts": "proceed",
    "source": {},
    "dest":{
      "op_type":"create"
    },
    "script": {
      "lang": "painless",
      "inline": ""
    }
  }
}
