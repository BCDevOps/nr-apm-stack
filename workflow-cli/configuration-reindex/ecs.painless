if (ctx._source.event.dataset == "apache.access") {
  if (ctx._source.service == null) {
    ctx._source.service = new HashMap();
  }
  if (ctx._source.offset != null) {
    ctx._source.event.sequence = ctx._source.remove("offset");
  }
  if (ctx._source.apache != null) {
    ctx._source.remove("apache");
  }
  if (ctx._source.message != null) {
    ctx._source.event.original = ctx._source.remove("message");
  }
  ctx._source.log.file.remove("extension");
  ctx._source.log.file.remove("name");
  ctx._source.log.file.remove("directory");
  ctx._source.remove("@event");
  if (ctx._source.url != null && ctx._source.url.file != null) {
    ctx._source.url.remove("file");
  }
  if (ctx._source.url != null && ctx._source.url.uri != null) {
    ctx._source.url.remove("uri");
  }
  if (ctx._source.url != null && ctx._source.url.query_v != null) {
    ctx._source.url.remove("query_v");
  }
  if (ctx._source.url != null && ctx._source.url.query_kv != null) {
    ctx._source.url.remove("query_kv");
  }
  if (ctx._source.url != null && ctx._source.url.query_k != null) {
    ctx._source.url.remove("query_k");
  }
  if (ctx._source.url != null && ctx._source.url.path_param != null) {
    ctx._source.url.remove("path_param");
  }
  if (ctx._source.labels.application != null) {
    ctx._source.labels.project = ctx._source.labels.remove("application");
  }
  if (ctx._source.labels.env_alias != null) {
    ctx._source.labels.environment_alias = ctx._source.labels.remove("env_alias");
  }
  if (ctx._source.labels.format != null) {
    ctx._source.labels.remove("format");
  }
  if (ctx._source.labels.context != null) {
    ctx._source.url.context = ctx._source.labels.remove("context");
  }
  if (ctx._source.labels.env != null) {
    ctx._source.service.environment = ctx._source.labels.remove("env");
    if (ctx._source.service.environment == "staging1" || ctx._source.service.environment == "staging2" || ctx._source.service.environment == "staging3") {
      ctx._source.service.environment = "staging";
    }
  }
  if (ctx._source.client != null) {
    ctx._source.source = ctx._source.remove("client");
  }
  if (ctx._source.http != null && ctx._source.http.request != null && ctx._source.http.request.referrer != null) {
    ctx._source.http.request.referrer = ctx._source.http.request.referrer.original;
  }
  if (ctx._source.http != null && ctx._source.http.request != null) {
    ctx._source.http.request.remove("line");
  }
  ctx._source.event.remove("original_timestamp");
  if (ctx._source.tls != null && ctx._source.tls.cypher != null) {
    ctx._source.tls.cipher = ctx._source.tls.remove("cypher");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.engine != null) {
    ctx._source.user_agent.remove("engine");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.cpu != null) {
    ctx._source.user_agent.remove("cpu");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.device != null && ctx._source.user_agent.device.name == null) {
    ctx._source.user_agent.remove("device");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.browser != null) {
    if (ctx._source.user_agent.browser.name != null && ctx._source.user_agent.name == null) {
      ctx._source.user_agent.name = ctx._source.user_agent.browser.name;
    }
    if (ctx._source.user_agent.browser.version != null && ctx._source.user_agent.version == null) {
      ctx._source.user_agent.version = ctx._source.user_agent.browser.version;
    }
    ctx._source.user_agent.remove("browser");
  }

  if (ctx._source.timestamp != null) {
    ctx._source.remove("timestamp");
  }

  if (ctx._source.event != null && ctx._source.event.kinesis != null) {
    ctx._source.kinesis = ctx._source.event.remove("kinesis");
  }
}
if (ctx._source.event.dataset == "express.access") {
  if (ctx._source.host.namespace != null) {
    ctx._source.host.name = ctx._source.host.remove("namespace");
  }

  if (ctx._source.http.request.referrer.uri != null) {
    ctx._source.http.request.referrer = ctx._source.http.request.referrer.uri;
  }

  if (ctx._source.http.response.time != null) {
    ctx._source.event.duration = ctx._source.http.response.remove("time") * 1000000;
  }
}

if (ctx._source.event.dataset == "node.css.http") {
  if (ctx._source.service == null) {
    ctx._source.service = new HashMap();
  }
  if (ctx._source.host == null) {
    ctx._source.host = new HashMap();
  }

  if (ctx._source.message != null) {
    ctx._source.event.original = ctx._source.remove("message");
  }
  if (ctx._source.timestamp != null) {
    ctx._source.remove("timestamp");
  }
  if (ctx._source.labels.env != null) {
    ctx._source.service.environment = ctx._source.labels.remove("env");
    if (ctx._source.service.environment == "staging1" || ctx._source.service.environment == "staging2" || ctx._source.service.environment == "staging3") {
      ctx._source.service.environment = "staging";
    }
  }
  if (ctx._source.labels.application != null) {
    ctx._source.labels.project = ctx._source.labels.remove("application");
    ctx._source.service.name = ctx._source.labels.project;
  }
  if (ctx._source.kubernetes != null) {
    if (ctx._source.kubernetes.namespace_name != null) {
      ctx._source.host.name = ctx._source.kubernetes.remove("namespace_name");
    }
    if (ctx._source.kubernetes.pod_ip != null) {
      ctx._source.host.ip = ctx._source.kubernetes.remove("pod_ip");
    }
    if (ctx._source.kubernetes.pod_name != null) {
      ctx._source.host.hostname = ctx._source.kubernetes.remove("pod_name");
    }
    ctx._source.remove("kubernetes");
  }

  if (ctx._source.log.file.offset != null) {
    ctx._source.event.sequence = ctx._source.log.file.remove("offset");
  }
  ctx._source.log.file.remove("extension");
  ctx._source.log.file.remove("directory");
  ctx._source.log.file.remove("name");
  ctx._source.remove("ip");
  ctx._source.event.type = "access";
  if (ctx._source.client != null) {
    ctx._source.source = ctx._source.remove("client");
  }
  if (ctx._source.http != null && ctx._source.http.response.time != null) {
    ctx._source.event.duration = ctx._source.http.response.remove("time") * 1000000;
  }
  if (ctx._source.http == null) {
    ctx['op'] = 'delete';
  }
}
