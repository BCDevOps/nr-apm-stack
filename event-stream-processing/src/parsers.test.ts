import {KinesisStreamEvent, KinesisStreamRecord} from 'aws-lambda';
import {create as buildContainer, TYPES} from './inversify.config';
import {KinesisStreamHandler} from './kinesisStreamHandler.isvc';
import {APACHE_ACCESS_LOG_EVENT_SIGNATURE} from './parser.apache.svc';
import {Randomizer} from './randomizer.isvc';
import * as lodash from 'lodash';
import {ParserEcs} from './parser.ecs.svc';
import {ParserApplicationClasification} from './parser.apps.svc';
import {Logger} from './logger.isvc';
import {LoggerVoidImpl} from './logger-void.svc';

const myContainer = buildContainer();

beforeEach(() => {
  myContainer.snapshot();
  myContainer.rebind<Randomizer>(TYPES.Randomizer).toConstantValue({randomBytes: ()=>{
    return Buffer.from([0x62, 0x75, 0x66, 0x66, 0x65, 0x72]);
  }});
  myContainer.rebind<Logger>(TYPES.Logger).to(LoggerVoidImpl);
});

afterEach(() => {
  myContainer.restore();
});

test('apache - safe uri', async () => {
  // eslint-disable-next-line max-len
  const message1 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [20/Apr/2021:15:10:40 -0700] "POST /int/fncs/activities/details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';

  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({...APACHE_ACCESS_LOG_EVENT_SIGNATURE, message: message1}), 'utf8',
          ).toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(1);
  for (const event of events) {
    expect(event).toHaveProperty('url.original');
    expect(event).toHaveProperty('url.path');
    expect(event).toHaveProperty('http.request.line');
    expect(event).toHaveProperty('http.request.method');
    expect(event).toHaveProperty('http.version');
    expect(event).toHaveProperty('event.ingested');
    delete event.event.ingested;
  }
  expect(events).toMatchSnapshot('8cb108e0-c707-4faf-83fc-10bc756f6b1b');
});

test('apache - safe-ish uri', async () => {
  // eslint-disable-next-line max-len
  const message1 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [20/Apr/2021:15:10:40 -0700] "POST /int/fncs/activities/ details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';
  // eslint-disable-next-line max-len
  const message2 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [20/Apr/2021:15:10:40 -0700] "POST /int/fncs/activities/%20details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';

  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({
              ...APACHE_ACCESS_LOG_EVENT_SIGNATURE,
              message: message1,
            }), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({
              ...APACHE_ACCESS_LOG_EVENT_SIGNATURE,
              message: message2,
            }), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(2);
  // expect(events).toMatchSnapshot('4a7f73e8-2b1b-43c9-babd-0a8deb1d09a2')
  for (const event of events) {
    expect(event).toHaveProperty('url.original');
    expect(event).toHaveProperty('url.path');
  }
});

test('apache - unsafe uri', async () => {
  // eslint-disable-next-line max-len
  const message1 = 'v1.0 20120211 "https://testapps.nrs.gov.bc.ca:443" "2001:569:be94:4700:61b4:917e:808:e3c6" [20/Apr/2021:15:10:40 -0700] "POST ../int/fncs/activities/details.xhtml HTTP/1.1" 200 2600 bytes 1112 bytes "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36" 133 ms, "TLSv1.2" "ECDHE-RSA-AES256-GCM-SHA384"';

  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({
              ...APACHE_ACCESS_LOG_EVENT_SIGNATURE,
              message: message1,
            }), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(1);
  for (const event of events) {
    expect(event).toHaveProperty('url.original');
    expect(event).not.toHaveProperty('url.path');
  }
  // expect(events).toMatchSnapshot('59d34900-3da1-4541-a3fc-2fcd43e2056c')
});


test('geoip - client.ip', async () => {
  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({
              ...APACHE_ACCESS_LOG_EVENT_SIGNATURE,
              client: {ip: '2001:569:be94:4700:61b4:917e:808:e3c6'},
            }), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(1);
  for (const event of events) {
    expect(event).toHaveProperty('client.geo.country_name');
    expect(event).toHaveProperty('client.geo.city_name');
    expect(event).toHaveProperty('client.geo.location');
    expect(event).toHaveProperty('event.ingested');
    delete event.event.ingested;
  }
  expect(events).toMatchSnapshot('eb9545ae-e9a8-41bb-8998-81fb0a1931b9');
});

test('geoip - source.address', async () => {
  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify({
              ...APACHE_ACCESS_LOG_EVENT_SIGNATURE,
              source: {address: '2001:569:be94:4700:61b4:917e:808:e3c6'},
            }), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(1);
  for (const event of events) {
    expect(event).toHaveProperty('source.geo.country_name');
    expect(event).toHaveProperty('source.geo.city_name');
    expect(event).toHaveProperty('source.geo.location');
  }
  // expect([]).toMatchSnapshot('59d34900-3da1-4541-a3fc-2fcd43e2056c')
});

test('apache - empty', async () => {
  const event:KinesisStreamEvent = {
    Records: [
            {
              kinesis: {
                data: Buffer.from(JSON.stringify({...APACHE_ACCESS_LOG_EVENT_SIGNATURE}), 'utf8').toString('base64'),
              },
            } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  for (const event of events) {
    expect(event).toHaveProperty('event.ingested');
    delete event.event.ingested;
  }
  expect(events).toHaveLength(1);
  expect(events).toMatchSnapshot('eb6509a2-5862-41fc-86a0-8f8e692a9131');
});

test('app labels', async () => {
  const event:KinesisStreamEvent = {
    Records: [
      {
        kinesis: {
          data: Buffer.from(
            JSON.stringify(
              lodash.merge({},
                APACHE_ACCESS_LOG_EVENT_SIGNATURE,
                {url: {domain: 'a100.gov.bc.ca', path: '/int/irs/hello'}}),
            ), 'utf8').toString('base64'),
        },
      } as any as KinesisStreamRecord,
    ],
  };
  const handler = myContainer.get<KinesisStreamHandler>(TYPES.KnesisStreamHandler);
  const events = await handler.transformToElasticCommonSchema(event);
  expect(events).toHaveLength(1);
  // await expect(events).toMatchSnapshot('fe9ed426-57e5-4148-ab8e-0dce6b2c517e')
  for (const event of events) {
    expect(event).toHaveProperty('labels.application');
    expect(event).toHaveProperty('labels.context');
    expect(event.labels.application).toEqual('irs');
  }
});

test('http event.outcome', () => {
  const parser = new ParserApplicationClasification();
  const record:any = lodash.merge(
    {},
    APACHE_ACCESS_LOG_EVENT_SIGNATURE,
    {url: {scheme: 'http', domain: 'sitesandtrailsbc.ca', path: '/welcome'}},
  );
  parser.apply(record);
  expect(record).toHaveProperty('event.ingested');
  delete record.ingested;
  expect(record).toMatchSnapshot('fe9ed426-57e5-4148-ab8e-0dce6b2c517e');
  expect(record).toHaveProperty('labels.application');
  // expect(record).toHaveProperty('labels.context')
  expect(record.labels.application).toEqual('sitesandtrailsbc');
  expect(record.labels).not.toHaveProperty('context');
});

test('app labels - sitesandtrailsbc.ca', () => {
  const parser = new ParserApplicationClasification();
  const record:any = lodash.merge(
    {},
    APACHE_ACCESS_LOG_EVENT_SIGNATURE,
    {url: {domain: 'sitesandtrailsbc.ca', path: '/welcome'}},
  );
  parser.apply(record);
  delete record.event.ingested;
  expect(record).toMatchSnapshot('fe9ed426-57e5-4148-ab8e-0dce6b2c517e');
  expect(record).toHaveProperty('labels.application');
  // expect(record).toHaveProperty('labels.context')
  expect(record.labels.application).toEqual('sitesandtrailsbc');
  expect(record.labels).not.toHaveProperty('context');
});

test('ecs - referrer - simple', () => {
  const parser = new ParserEcs();
  const record = {http: {request: {referrer: {original: 'http://somewhere.com/with/some/path.ext'}}}};
  parser.apply(record);
  expect(record).toHaveProperty('http.request.referrer.original');
  expect(record).toHaveProperty('http.request.referrer.scheme');
  expect(record).toHaveProperty('http.request.referrer.domain');
  expect(record).toHaveProperty('http.request.referrer.port');
  expect(record).toHaveProperty('http.request.referrer.path');
  expect(record).not.toHaveProperty('http.request.referrer.username');
  expect(record).not.toHaveProperty('http.request.referrer.query');
  expect(record).not.toHaveProperty('http.request.referrer.fragment');
  // expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - no path', () => {
  const parser = new ParserEcs();
  const record: any = {http: {request: {referrer: {original: 'http://somewhere.com'}}}};
  const matches = parser.matches();
  parser.apply(record);
  expect(matches).toEqual(true);
  expect(record).toHaveProperty('http.request.referrer.original');
  expect(record).toHaveProperty('http.request.referrer.scheme');
  expect(record).toHaveProperty('http.request.referrer.domain');
  expect(record).toHaveProperty('http.request.referrer.port');
  expect(record).toHaveProperty('http.request.referrer.path');
  expect(record.http.request.referrer.path).toMatch(/^\//);
  expect(record).not.toHaveProperty('http.request.referrer.username');
  expect(record).not.toHaveProperty('http.request.referrer.query');
  expect(record).not.toHaveProperty('http.request.referrer.fragment');
  expect(record).toMatchSnapshot('fe3a90a4-1319-44e0-af0f-312686d81ae7');
});

test('ecs - referrer - path with just /', () => {
  const parser = new ParserEcs();
  const record:any = {http: {request: {referrer: {original: 'http://somewhere.com/'}}}};
  parser.apply(record);
  expect(record).toHaveProperty('http.request.referrer.original');
  expect(record).toHaveProperty('http.request.referrer.scheme');
  expect(record).toHaveProperty('http.request.referrer.domain');
  expect(record).toHaveProperty('http.request.referrer.port');
  expect(record.http.request.referrer.port).toEqual('80');
  expect(record).toHaveProperty('http.request.referrer.path');
  expect(record.http.request.referrer.path).toMatch(/^\//);
  expect(record).not.toHaveProperty('http.request.referrer.username');
  expect(record).not.toHaveProperty('http.request.referrer.query');
  expect(record).not.toHaveProperty('http.request.referrer.fragment');
  expect(record).toMatchSnapshot('84023bc4-77fe-439f-9123-91c20362c04a');
});
