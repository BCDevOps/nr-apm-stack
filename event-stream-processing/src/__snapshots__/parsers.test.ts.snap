// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apache - empty: eb6509a2-5862-41fc-86a0-8f8e692a9131 1`] = `
Array [
  Object {
    "_id": "756e646566696e65642e363237353636363636353732",
    "event": Object {
      "category": "web",
      "dataset": "apache.access",
      "kind": "event",
      "risk_score": 1,
    },
  },
]
`;

exports[`apache - safe uri: 8cb108e0-c707-4faf-83fc-10bc756f6b1b 1`] = `
Array [
  Object {
    "@timestamp": "2021-04-20T15:10:40.000-07:00",
    "_id": "756e646566696e65642e363237353636363636353732",
    "_index": "nrm-logs-access-2021.04.20",
    "apache": Object {
      "version": "20120211",
    },
    "client": Object {
      "as": Object {
        "number": 852,
        "organization": Object {
          "name": "ASN852",
        },
      },
      "geo": Object {
        "city_name": "Victoria",
        "continent_code": "NA",
        "continent_name": "North America",
        "country_iso_code": "CA",
        "country_name": "Canada",
        "location": Object {
          "lat": 48.4771,
          "lon": -123.5313,
        },
        "postal_code": "V9B",
        "region_iso_code": "BC",
        "region_name": "British Columbia",
        "timezone": "America/Vancouver",
      },
      "ip": "2001:569:be94:4700:61b4:917e:808:e3c6",
    },
    "event": Object {
      "category": "web",
      "dataset": "apache.access",
      "duration": "133",
      "kind": "event",
      "outcome": "success",
    },
    "http": Object {
      "request": Object {
        "bytes": "2600",
        "line": "POST /int/fncs/activities/details.xhtml HTTP/1.1",
        "method": "POST",
        "referrer": Object {
          "domain": "testapps.nrs.gov.bc.ca",
          "file": Object {
            "ext": "xhtml",
            "name": "details.xhtml",
          },
          "full": "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC",
          "original": "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC",
          "path": "/int/fncs/activities/details.xhtml",
          "port": "443",
          "query": "activityGuid=0123C676CBAB461E8C6A0A14567A16AC",
          "query_k": Array [
            "activityGuid",
          ],
          "query_kv": Array [
            "activityGuid=0123C676CBAB461E8C6A0A14567A16AC",
          ],
          "query_v": Array [
            "0123C676CBAB461E8C6A0A14567A16AC",
          ],
          "scheme": "https",
          "uri": "//testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml",
        },
      },
      "response": Object {
        "bytes": "1112",
        "status_code": "200",
      },
      "version": "1.1",
    },
    "labels": Object {
      "application": "fncs",
      "context": "/int",
      "format": "v1.0",
    },
    "message": "v1.0 20120211 \\"https://testapps.nrs.gov.bc.ca:443\\" \\"2001:569:be94:4700:61b4:917e:808:e3c6\\" [20/Apr/2021:15:10:40 -0700] \\"POST /int/fncs/activities/details.xhtml HTTP/1.1\\" 200 2600 bytes 1112 bytes \\"https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml?activityGuid=0123C676CBAB461E8C6A0A14567A16AC\\" \\"Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36\\" 133 ms, \\"TLSv1.2\\" \\"ECDHE-RSA-AES256-GCM-SHA384\\"",
    "tls": Object {
      "cypher": "ECDHE-RSA-AES256-GCM-SHA384",
      "version_protocol": "TLSv1.2",
    },
    "url": Object {
      "domain": "testapps.nrs.gov.bc.ca",
      "file": Object {
        "ext": "xhtml",
        "name": "details.xhtml",
      },
      "full": "https://testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml",
      "original": "/int/fncs/activities/details.xhtml",
      "path": "/int/fncs/activities/details.xhtml",
      "port": "443",
      "scheme": "https",
      "uri": "//testapps.nrs.gov.bc.ca/int/fncs/activities/details.xhtml",
    },
    "user_agent": Object {
      "name": "Chrome",
      "original": "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.128 Safari/537.36",
      "os": Object {
        "name": "Windows",
        "version": "10",
      },
      "version": "89.0.4389.128",
    },
  },
]
`;

exports[`app labels - sitesandtrailsbc.ca: fe9ed426-57e5-4148-ab8e-0dce6b2c517e 1`] = `
Object {
  "event": Object {
    "category": "web",
    "dataset": "apache.access",
    "kind": "event",
  },
  "labels": Object {
    "application": "sitesandtrailsbc",
  },
  "url": Object {
    "domain": "sitesandtrailsbc.ca",
    "path": "/welcome",
  },
}
`;

exports[`ecs - referrer - no path: fe3a90a4-1319-44e0-af0f-312686d81ae7 1`] = `
Object {
  "http": Object {
    "request": Object {
      "referrer": Object {
        "domain": "somewhere.com",
        "full": "http://somewhere.com/",
        "original": "http://somewhere.com",
        "path": "/",
        "port": "80",
        "scheme": "http",
        "uri": "//somewhere.com/",
      },
    },
  },
}
`;

exports[`ecs - referrer - path with just /: 84023bc4-77fe-439f-9123-91c20362c04a 1`] = `
Object {
  "http": Object {
    "request": Object {
      "referrer": Object {
        "domain": "somewhere.com",
        "full": "http://somewhere.com/",
        "original": "http://somewhere.com/",
        "path": "/",
        "port": "80",
        "scheme": "http",
        "uri": "//somewhere.com/",
      },
    },
  },
}
`;

exports[`geoip - client.ip: eb9545ae-e9a8-41bb-8998-81fb0a1931b9 1`] = `
Array [
  Object {
    "_id": "756e646566696e65642e363237353636363636353732",
    "client": Object {
      "as": Object {
        "number": 852,
        "organization": Object {
          "name": "ASN852",
        },
      },
      "geo": Object {
        "city_name": "Victoria",
        "continent_code": "NA",
        "continent_name": "North America",
        "country_iso_code": "CA",
        "country_name": "Canada",
        "location": Object {
          "lat": 48.4771,
          "lon": -123.5313,
        },
        "postal_code": "V9B",
        "region_iso_code": "BC",
        "region_name": "British Columbia",
        "timezone": "America/Vancouver",
      },
      "ip": "2001:569:be94:4700:61b4:917e:808:e3c6",
    },
    "event": Object {
      "category": "web",
      "dataset": "apache.access",
      "kind": "event",
      "risk_score": 1,
    },
  },
]
`;

exports[`http event.outcome: fe9ed426-57e5-4148-ab8e-0dce6b2c517e 1`] = `
Object {
  "event": Object {
    "category": "web",
    "dataset": "apache.access",
    "ingested": "2021-05-26T18:47:40.314-07:00",
    "kind": "event",
  },
  "labels": Object {
    "application": "sitesandtrailsbc",
  },
  "url": Object {
    "domain": "sitesandtrailsbc.ca",
    "path": "/welcome",
    "scheme": "http",
  },
}
`;
